pkg/util/tracingservice/nodes.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tracingservice
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/redact"
+)
+
+// node captures the relevant bits of each node as it pertains to the tracing
+// service infrastructure.
+type node struct {
+	id    roachpb.NodeID
+	epoch int64
+}
+
+// nodes is a collection of node objects.
+type nodes []node
+
+// nodesFromNodeLiveness returns the IDs and epochs for all nodes that are
+// currently part of the cluster (i.e. they haven't been decommissioned away).
+// If any of the nodes are found to be unavailable, an error is returned.
+//
+// It's important to note that this makes no guarantees about new nodes being
+// added to the cluster. It's entirely possible for that to happen concurrently
+// with the retrieval of the current set of nodes.
+func nodesFromNodeLiveness(ctx context.Context, nl NodeLiveness) (nodes, error) {
+	var ns []node
+	ls, err := nl.GetLivenessesFromKV(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, l := range ls {
+		if l.Membership.Decommissioned() {
+			continue
+		}
+		live, err := nl.IsLive(l.NodeID)
+		if err != nil {
+			return nil, err
+		}
+		if !live {
```

Comment:
I don't think you care about this for observability. If a node is not live, do we want to error out or do you want to skip over it (though maybe mentioning somewhere that you did)? Even then, we'd time out if issuing RPCs to dead nodes. This filter here is basically making sure we don't even try. Perhaps that's what we want, but timing out would give us a convenient place to capture the error and propagate it somehow -- if that's what we want.

Reply:
I was thinking we'd just use node liveness to retrieve a list of non-decommissioned nodes.

Reply:
good point, changed to only returning non-decommissioned nodes. Also added a comment explaining that we will rely on the RPC timing out so we can surface something along the lines of "trace for node x is missing".